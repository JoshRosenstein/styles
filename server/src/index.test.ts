/**
 * @jest-environment node
 */
import * as fs from 'fs'
import styles from '../../src'
import {
  createStyleTagFromString,
  createStyleTagFromCache,
  writeStylesFromCache,
  writeStylesFromString,
} from './index'

describe('Configure', () => {
  it('removes vendor prefixing', () => {
    const myStyles = styles.create({prefix: false})
    const style = myStyles({
      flex: {display: 'flex'},
    })

    style('flex')
    expect(createStyleTagFromCache(myStyles)).toMatchSnapshot()
  })

  it('has customized vendor prefixing', () => {
    // adds prefixes to transform, but not flex
    const prefix = function (key, value, context): boolean {
      if (typeof context !== 'number') throw 'fail'

      switch (key) {
        case 'transform':
          return true
        // @ts-ignore
        case 'disable':
          if (value !== 'flex') throw 'fail'
        // eslint-disable-next-line
        default:
          return false
      }
    }

    const myStyles = styles.create({prefix})
    const style = myStyles({
      flex: {display: 'flex', transform: 'translateX(30px)'},
    })

    style('flex')
    expect(createStyleTagFromCache(myStyles)).toMatchSnapshot()
  })

  it('changes key to "css"', () => {
    const myStyles = styles.create({key: 'css'})
    const style = myStyles({
      flex: {display: 'flex'},
    })

    style('flex')
    expect(createStyleTagFromCache(myStyles)).toMatchSnapshot()
  })
})

describe('createStyleTagFromCache', () => {
  it('adds nonce to style tags', () => {
    const myStyles = styles.create({nonce: 'EDNnf03nceIOfn39fn3e9h3sdfa'})
    const style = myStyles({
      flex: {display: 'flex'},
    })

    style('flex')
    expect(createStyleTagFromCache(myStyles)).toMatchSnapshot()
  })

  it('extracts global styles', () => {
    const myStyles = styles.create({})
    myStyles.global`
      :root {
        --hello: "world";
      } 
    `

    expect(createStyleTagFromCache(myStyles)).toMatchSnapshot()
  })

  it('extracts global variables', () => {
    const myStyles = styles.create({})
    myStyles.variables({
      colors: {
        blue: '#09a',
      },
    })

    expect(createStyleTagFromCache(myStyles)).toMatchSnapshot()
  })

  it('extracts theme variables', () => {
    const myStyles = styles.create({})
    myStyles.themes({
      dark: {
        colors: {
          primary: '#000',
        },
      },
      light: {
        colors: {
          primary: '#fff',
        },
      },
    })

    expect(createStyleTagFromCache(myStyles)).toMatchSnapshot()
    myStyles.theme('dark')
    expect(createStyleTagFromCache(myStyles)).toMatchSnapshot()
    myStyles.theme('light')
    expect(createStyleTagFromCache(myStyles)).toMatchSnapshot()
  })

  it('caches styles', () => {
    const myStyles = styles.create({})
    const style = myStyles({
      flex: {display: 'flex'},
      btn: `
        border-radius: 1000px;
        background: blue;
        color: white;
      `,
    })

    style('flex')
    style('btn')
    style('flex')
    style('btn')
    expect(createStyleTagFromCache(myStyles)).toMatchSnapshot()
  })
})

describe('createStyleTagFromString', () => {
  it('adds nonce to style tags', () => {
    const myStyles = styles.create({nonce: 'EDNnf03nceIOfn39fn3e9h3sdfa'})
    const style = myStyles({
      flex: {display: 'flex'},
    })

    style('flex')
    expect(
      createStyleTagFromString(`<div className=${style('flex')}>`, myStyles)
    ).toMatchSnapshot()
  })

  it('extracts global styles', () => {
    const myStyles = styles.create({})
    myStyles.global`
      :root {
        --hello: "world";
      }
    `

    expect(createStyleTagFromString('', myStyles)).toMatchSnapshot()
  })

  it('extracts global variables', () => {
    const myStyles = styles.create({})
    myStyles.variables({
      colors: {
        blue: '#09a',
      },
    })

    expect(createStyleTagFromString('', myStyles)).toMatchSnapshot()
  })

  it('caches styles', () => {
    const myStyles = styles.create({})
    const style = myStyles({
      flex: {display: 'flex'},
      btn: `
        border-radius: 1000px;
        background: blue;
        color: white;
      `,
    })

    style('flex')
    style('btn')
    style('flex')
    style('btn')
    expect(
      createStyleTagFromString(
        `
      <div class=${style('flex')}>
        <div class=${style('btn')}>
          Hello
        </div>
      </div>
    `,
        myStyles
      )
    ).toMatchSnapshot()
  })
})

describe('writeStylesFromString', () => {
  it('writes', async () => {
    const myStyles = styles.create({})
    const style = myStyles({
      flex: {display: 'flex'},
      btn: `
        border-radius: 1000px;
        background: blue;
        color: white;
      `,
    })

    const finfo = await writeStylesFromString(
      `
        <div class=${style('flex')}>
          <div class=${style('btn')}>
            Hello
          </div>
        </div>
      `,
      './',
      myStyles
    )
    expect(fs.existsSync(finfo.filename)).toBe(true)
    fs.unlinkSync(finfo.filename)
    expect(finfo).toMatchSnapshot()
  })

  it('writes custom name', async () => {
    const myStyles = styles.create({})
    const style = myStyles({
      flex: {display: 'flex'},
      btn: `
        border-radius: 1000px;
        background: blue;
        color: white;
      `,
    })
    style('flex')
    style('btn')

    const finfo = await writeStylesFromString(
      `
        <div class=${style('flex')}>
          <div class=${style('btn')}>
            Hello
          </div>
        </div>
      `,
      './',
      myStyles,
      {name: 'foo.css'}
    )
    expect(fs.existsSync(finfo.filename)).toBe(true)
    fs.unlinkSync(finfo.filename)
    expect(finfo).toMatchSnapshot()
  })

  it('writes custom key', async () => {
    const myStyles = styles.create({key: 'css'})
    const style = myStyles({
      flex: {display: 'flex'},
      btn: `
        border-radius: 1000px;
        background: blue;
        color: white;
      `,
    })
    style('flex')
    style('btn')

    const finfo = await writeStylesFromString(
      `
        <div class=${style('flex')}>
          <div class=${style('btn')}>
            Hello
          </div>
        </div>
      `,
      './',
      myStyles
    )
    expect(fs.existsSync(finfo.filename)).toBe(true)
    fs.unlinkSync(finfo.filename)
    expect(finfo).toMatchSnapshot()
  })

  it('writes custom hash', async () => {
    const myStyles = styles.create({hash: () => 'f8bCooDawg'})
    const style = myStyles({
      flex: {display: 'flex'},
      btn: `
        border-radius: 1000px;
        background: blue;
        color: white;
      `,
    })
    style('flex')
    style('btn')

    const finfo = await writeStylesFromString(
      `
        <div class=${style('flex')}>
          <div class=${style('btn')}>
            Hello
          </div>
        </div>
      `,
      './',
      myStyles
    )
    expect(fs.existsSync(finfo.filename)).toBe(true)
    fs.unlinkSync(finfo.filename)
    expect(finfo).toMatchSnapshot()
  })
})

describe('writeStylesFromCache', () => {
  it('writes', async () => {
    const myStyles = styles.create({})
    const style = myStyles({
      flex: {display: 'flex'},
      btn: `
        border-radius: 1000px;
        background: blue;
        color: white;
      `,
    })
    style('flex')
    style('btn')

    const finfo = await writeStylesFromCache('./', myStyles)
    expect(fs.existsSync(finfo.filename)).toBe(true)
    fs.unlinkSync(finfo.filename)
    expect(finfo).toMatchSnapshot()
  })

  it('writes custom name', async () => {
    const myStyles = styles.create({})
    const style = myStyles({
      flex: {display: 'flex'},
      btn: `
        border-radius: 1000px;
        background: blue;
        color: white;
      `,
    })
    style('flex')
    style('btn')

    const finfo = await writeStylesFromCache('./', myStyles, {name: 'foo.css'})
    expect(fs.existsSync(finfo.filename)).toBe(true)
    fs.unlinkSync(finfo.filename)
    expect(finfo).toMatchSnapshot()
  })

  it('writes custom key', async () => {
    const myStyles = styles.create({key: 'css'})
    const style = myStyles({
      flex: {display: 'flex'},
      btn: `
        border-radius: 1000px;
        background: blue;
        color: white;
      `,
    })
    style('flex')
    style('btn')

    const finfo = await writeStylesFromCache('./', myStyles)
    expect(fs.existsSync(finfo.filename)).toBe(true)
    fs.unlinkSync(finfo.filename)
    expect(finfo).toMatchSnapshot()
  })

  it('writes custom hash', async () => {
    const myStyles = styles.create({hash: () => 'f8bCooDawg'})
    const style = myStyles({
      flex: {display: 'flex'},
      btn: `
        border-radius: 1000px;
        background: blue;
        color: white;
      `,
    })
    style('flex')
    style('btn')

    const finfo = await writeStylesFromCache('./', myStyles)
    expect(fs.existsSync(finfo.filename)).toBe(true)
    fs.unlinkSync(finfo.filename)
    expect(finfo).toMatchSnapshot()
  })
})
