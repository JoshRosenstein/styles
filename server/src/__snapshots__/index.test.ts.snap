// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Configure changes key to "css" 1`] = `"<style data-dash=\\"1ut9bc3\\" data-cache=\\"css\\">.css-1ut9bc3{display:-webkit-box;display:-webkit-flex;display:-ms-flexbox;display:flex;}</style>"`;

exports[`Configure has customized vendor prefixing 1`] = `"<style data-dash=\\"1vlpa1m\\" data-cache=\\"ui\\">.ui-1vlpa1m{display:flex;-webkit-transform:translateX(30px);-ms-transform:translateX(30px);transform:translateX(30px);}</style>"`;

exports[`Configure removes vendor prefixing 1`] = `"<style data-dash=\\"1ut9bc3\\" data-cache=\\"ui\\">.ui-1ut9bc3{display:flex;}</style>"`;

exports[`createStyleTagFromCache adds nonce to style tags 1`] = `"<style data-dash=\\"1ut9bc3\\" data-cache=\\"ui\\" nonce=\\"EDNnf03nceIOfn39fn3e9h3sdfa\\">.ui-1ut9bc3{display:-webkit-box;display:-webkit-flex;display:-ms-flexbox;display:flex;}</style>"`;

exports[`createStyleTagFromCache caches styles 1`] = `"<style data-dash=\\"1ut9bc3 1h3n8qo\\" data-cache=\\"ui\\">.ui-1ut9bc3{display:-webkit-box;display:-webkit-flex;display:-ms-flexbox;display:flex;}.ui-1h3n8qo{border-radius:1000px;background:blue;color:white;}</style>"`;

exports[`createStyleTagFromCache clears cached styles after render when \`clearCache\` is true 1`] = `"<style data-dash=\\"1ut9bc3 1h3n8qo\\" data-cache=\\"ui\\">.ui-1ut9bc3{display:-webkit-box;display:-webkit-flex;display:-ms-flexbox;display:flex;}.ui-1h3n8qo{border-radius:1000px;background:blue;color:white;}</style>"`;

exports[`createStyleTagFromCache clears cached styles after render when \`clearCache\` is true: empty 1`] = `"<style data-dash=\\"\\" data-cache=\\"ui\\"></style>"`;

exports[`createStyleTagFromCache extracts global styles 1`] = `"<style data-dash=\\"pf2fa2\\" data-cache=\\"ui\\">:root{--hello:\\"world\\";}</style>"`;

exports[`createStyleTagFromCache extracts global variables 1`] = `"<style data-dash=\\"1ahrdya\\" data-cache=\\"ui\\">:root{--colors-blue:#09a;}</style>"`;

exports[`createStyleTagFromCache extracts theme variables 1`] = `"<style data-dash=\\"1uitam7 14zev9x\\" data-cache=\\"ui\\">.ui-dark-theme{--colors-primary:#000;}.ui-light-theme{--colors-primary:#fff;}</style>"`;

exports[`createStyleTagFromCache extracts theme variables 2`] = `"<style data-dash=\\"1uitam7 14zev9x\\" data-cache=\\"ui\\">.ui-dark-theme{--colors-primary:#000;}.ui-light-theme{--colors-primary:#fff;}</style>"`;

exports[`createStyleTagFromCache extracts theme variables 3`] = `"<style data-dash=\\"1uitam7 14zev9x\\" data-cache=\\"ui\\">.ui-dark-theme{--colors-primary:#000;}.ui-light-theme{--colors-primary:#fff;}</style>"`;

exports[`createStyleTagFromCache uses default styles 1`] = `"<style data-dash=\\"1ut9bc3\\" data-cache=\\"ui\\">.ui-1ut9bc3{display:-webkit-box;display:-webkit-flex;display:-ms-flexbox;display:flex;}</style>"`;

exports[`createStyleTagFromString adds nonce to style tags 1`] = `"<style data-dash=\\"1ut9bc3\\" data-cache=\\"ui\\" nonce=\\"EDNnf03nceIOfn39fn3e9h3sdfa\\">.ui-1ut9bc3{display:-webkit-box;display:-webkit-flex;display:-ms-flexbox;display:flex;}</style>"`;

exports[`createStyleTagFromString caches styles 1`] = `"<style data-dash=\\"1ut9bc3 1h3n8qo\\" data-cache=\\"ui\\">.ui-1ut9bc3{display:-webkit-box;display:-webkit-flex;display:-ms-flexbox;display:flex;}.ui-1h3n8qo{border-radius:1000px;background:blue;color:white;}</style>"`;

exports[`createStyleTagFromString extracts global styles 1`] = `"<style data-dash=\\"pf2fa2\\" data-cache=\\"ui\\">:root{--hello:\\"world\\";}</style>"`;

exports[`createStyleTagFromString extracts global variables 1`] = `"<style data-dash=\\"1ahrdya\\" data-cache=\\"ui\\">:root{--colors-blue:#09a;}</style>"`;

exports[`createStyleTagFromString uses default styles instance 1`] = `"<style data-dash=\\"1ut9bc3\\" data-cache=\\"ui\\">.ui-1ut9bc3{display:-webkit-box;display:-webkit-flex;display:-ms-flexbox;display:flex;}</style>"`;

exports[`writeStylesFromCache writes 1`] = `
Object {
  "css": ".ui-1ut9bc3{display:-webkit-box;display:-webkit-flex;display:-ms-flexbox;display:flex;}.ui-1h3n8qo{border-radius:1000px;background:blue;color:white;}",
  "filename": "ui-ws4vpy.css",
  "name": "ui-ws4vpy.css",
  "names": Array [
    "1ut9bc3",
    "1h3n8qo",
  ],
  "path": "./",
}
`;

exports[`writeStylesFromCache writes custom hash 1`] = `
Object {
  "css": ".ui-f8bCooDawg{display:-webkit-box;display:-webkit-flex;display:-ms-flexbox;display:flex;}",
  "filename": "ui-f8bCooDawg.css",
  "name": "ui-f8bCooDawg.css",
  "names": Array [
    "f8bCooDawg",
  ],
  "path": "./",
}
`;

exports[`writeStylesFromCache writes custom key 1`] = `
Object {
  "css": ".css-1ut9bc3{display:-webkit-box;display:-webkit-flex;display:-ms-flexbox;display:flex;}.css-1h3n8qo{border-radius:1000px;background:blue;color:white;}",
  "filename": "css-13ed95k.css",
  "name": "css-13ed95k.css",
  "names": Array [
    "1ut9bc3",
    "1h3n8qo",
  ],
  "path": "./",
}
`;

exports[`writeStylesFromCache writes custom name 1`] = `
Object {
  "css": ".ui-1ut9bc3{display:-webkit-box;display:-webkit-flex;display:-ms-flexbox;display:flex;}.ui-1h3n8qo{border-radius:1000px;background:blue;color:white;}",
  "filename": "foo.css",
  "name": "foo.css",
  "names": Array [
    "1ut9bc3",
    "1h3n8qo",
  ],
  "path": "./",
}
`;

exports[`writeStylesFromCache writes from default styles instance 1`] = `
Object {
  "css": ".ui-1ut9bc3{display:-webkit-box;display:-webkit-flex;display:-ms-flexbox;display:flex;}.ui-1h3n8qo{border-radius:1000px;background:blue;color:white;}",
  "filename": "ui-ws4vpy.css",
  "name": "ui-ws4vpy.css",
  "names": Array [
    "1ut9bc3",
    "1h3n8qo",
  ],
  "path": "",
}
`;

exports[`writeStylesFromString writes 1`] = `
Object {
  "css": ".ui-1ut9bc3{display:-webkit-box;display:-webkit-flex;display:-ms-flexbox;display:flex;}.ui-1h3n8qo{border-radius:1000px;background:blue;color:white;}",
  "filename": "ui-ws4vpy.css",
  "name": "ui-ws4vpy.css",
  "names": Array [
    "1ut9bc3",
    "1h3n8qo",
  ],
  "path": "./",
}
`;

exports[`writeStylesFromString writes custom hash 1`] = `
Object {
  "css": ".ui-f8bCooDawg{display:-webkit-box;display:-webkit-flex;display:-ms-flexbox;display:flex;}",
  "filename": "ui-f8bCooDawg.css",
  "name": "ui-f8bCooDawg.css",
  "names": Array [
    "f8bCooDawg",
  ],
  "path": "./",
}
`;

exports[`writeStylesFromString writes custom key 1`] = `
Object {
  "css": ".css-1ut9bc3{display:-webkit-box;display:-webkit-flex;display:-ms-flexbox;display:flex;}.css-1h3n8qo{border-radius:1000px;background:blue;color:white;}",
  "filename": "css-13ed95k.css",
  "name": "css-13ed95k.css",
  "names": Array [
    "1ut9bc3",
    "1h3n8qo",
  ],
  "path": "./",
}
`;

exports[`writeStylesFromString writes custom name 1`] = `
Object {
  "css": ".ui-1ut9bc3{display:-webkit-box;display:-webkit-flex;display:-ms-flexbox;display:flex;}.ui-1h3n8qo{border-radius:1000px;background:blue;color:white;}",
  "filename": "foo.css",
  "name": "foo.css",
  "names": Array [
    "1ut9bc3",
    "1h3n8qo",
  ],
  "path": "./",
}
`;

exports[`writeStylesFromString writes from default styles instance 1`] = `
Object {
  "css": ".ui-1ut9bc3{display:-webkit-box;display:-webkit-flex;display:-ms-flexbox;display:flex;}.ui-1h3n8qo{border-radius:1000px;background:blue;color:white;}",
  "filename": "ui-ws4vpy.css",
  "name": "ui-ws4vpy.css",
  "names": Array [
    "1ut9bc3",
    "1h3n8qo",
  ],
  "path": "./",
}
`;
